opwen_email_server/mailers/shared.py

```diff
@@ -0,0 +1,47 @@
+from typing import Callable
+from typing import Dict
+
+from libcloud.storage.types import ObjectDoesNotExistError
+
+from opwen_email_server.actions import _Action
+from opwen_email_server.services.storage import AzureObjectStorage
+from opwen_email_server.services.storage import AzureTextStorage
+from opwen_email_server.utils.email_parser import MimeEmailParser
+from opwen_email_server.utils.unique import new_email_id
+
+
+class ProcessServiceEmail(_Action):
+    def __init__(self,
+                 raw_email_storage: AzureTextStorage,
+                 email_storage: AzureObjectStorage,
+                 next_task: Callable[[str], None],
+                 registry: Dict[str, Callable[[dict], dict]],
+                 email_parser: Callable[[dict], dict] = None):
+
+        self._raw_email_storage = raw_email_storage
+        self._email_storage = email_storage
+        self._next_task = next_task
+        self._registry = registry
+        self._email_parser = email_parser or MimeEmailParser()
+
+    def _action(self, resource_id):  # type: ignore
+        try:
+            mime_email = self._raw_email_storage.fetch_text(resource_id)
+        except ObjectDoesNotExistError:
+            self.log_warning('Inbound email %s does not exist', resource_id)
+            return 'skipped', 202
+
+        email = self._email_parser(mime_email)
+
+        for address in email['to']:
+            mailer_service = self._registry[address]
+            formatted_email = mailer_service(email)
+
+            email_id = new_email_id(formatted_email)
```

Comment:
Just for consistency, I'd rename this `formatted_email_id` so it's explicit that it's the id for the new email we just created. When initially reading this I did a little double-take asking myself _"why are we storing the email in the loop? new_email_id will always return the same value given the same email"_ and then realized that we're actually operating on the formatted email (where of course the behavior implemented here is correct). Renaming will prevent this situation when someone is only skimming the code quickly.

Reply:
Ok, will do! Making things easier to understand for skimming purposes is always a good thing!